### ISSUE TITLE
context information pane messes up objects

### ISSUE THREAD

**Issue by dirstel** [node-red/node-red#2809]
### What are the steps to reproduce?
* use filesystem to store context data (maybe this is not necessary)
![image](https://user-images.githubusercontent.com/24778060/103612752-61e95d00-4f25-11eb-9be9-6961d13dcec0.png)

* use a change-node to store an object in flow context
![image](https://user-images.githubusercontent.com/24778060/103612929-c5738a80-4f25-11eb-8081-58afd3abc457.png)

* use a change-node to assign this object to a payload.property
![image](https://user-images.githubusercontent.com/24778060/103612808-847b7600-4f25-11eb-823b-90bbbccb66f4.png)

* use a change-node to add more properties to the payload.property
![image](https://user-images.githubusercontent.com/24778060/103612819-8cd3b100-4f25-11eb-8f6d-de47465059ec.png)

### What happens?
* changes applied to payload.property are reflected as part of the context data object in the context information pane
![image](https://user-images.githubusercontent.com/24778060/103612866-9f4dea80-4f25-11eb-89a2-1d6945b733d8.png)

* this is not applied to storage of context data
![image](https://user-images.githubusercontent.com/24778060/103612873-a2e17180-4f25-11eb-8fff-445c5586fbf5.png)

### What do you expect to happen?
* Assigning an object to an payload property should create a new and independant copy of object. Changes applliued here should not be seen in origin object.

### Please tell us about your environment:

- [x] Node-RED version: 1.2.5 (offical Docker Image)
- [x] Node.js version: (offical Docker Image)
- [x] npm version: (offical Docker Image)
- [x] Platform/OS: Ubuntu 18.04.3 / Docker 18.06.0-ce
- [x] Browser: Chrome 87.0.4280.88

**Comment by knolleary** [comment]
The context store has not been updated because no explicit call to set the new value in the store has been made.

We don't clone the objects returned from context as in many cases it is not necessary to do so and would cause a significant performance impact if we did.

That can lead to unexpected results as you've observed here.

When using the Function node for example, you can clone the object returned from Context to prevent any accidental modification to the object. But the Change node doesn't provide a way to conveniently do that.

Given this does catch users out, we do need to do something - the question is what that is. We will not start cloning everything returned from context due to the performance impact that would have to all users who are not otherwise affected by this behaviour.

Some options to consider are:
 1. Add a configuration option to the Context Store in settings.js to enable cloning. That will have limited benefit as users won't easily discover that option and know to turn it on.
 2. Update the Context store API for the `get` function to add a flag to enabling cloning for that call. This would be more convenient for the user, although would require some careful design work to expose in the TypedInput widget used by the Change node etc.

**Comment by dirstel** [comment]
Some further options:

3. Add Checkbox "clone" to change-node (clone only if needed)
4. clone in change node by default (breaks default behaviour)
5. add config to switch cloning for change-node on/off (may lead to different behaviour in different installations and somewhat hard bugtracking)

**Comment by tve** [comment]
I'm seeing a problem that may be related: the context information pane messes up for keys that contain a period (the original issue post has periods in the key).

To reproduce:
- create a function node with the following code:
```
global.set('foo.bar', {this: 'that'})
return {payload: JSON.stringify(global.get('foo.bar'))}
```
- use an inject node to trigger the function node
- refresh the global context in the context side-panel
- observe a messed-up display (the key should be foo.bar, and bar should not show up in the value):
![image](https://user-images.githubusercontent.com/39480/156815709-99963edf-2b4c-42e1-a410-b7dcaf45203c.png)
- attach a debug node to the output of the function node to print the msg.payload
- observe the correct output:
![image](https://user-images.githubusercontent.com/39480/156815838-9a8ad6a1-6625-4d67-a4c5-06de87ba0295.png)

Here's a flow with the three mentioned nodes:
```
[{"id":"e6f818616e6b9542","type":"function","z":"56494f06ecf796f0","name":"","func":"global.set('foo.bar', {this: 'that'})\nreturn {payload: JSON.stringify(global.get('foo.bar'))}","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":340,"y":460,"wires":[["0e9ab6660f1e5dad"]]},{"id":"1856ac79b13f46c6","type":"inject","z":"56494f06ecf796f0","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":180,"y":460,"wires":[["e6f818616e6b9542"]]},{"id":"0e9ab6660f1e5dad","type":"debug","z":"56494f06ecf796f0","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":510,"y":460,"wires":[]}]
```

**Comment by knolleary** [comment]
Nope - that is working as designed. Keys are treated as javascript property expressions. So if you use a key of `foo.bar` that is saying create a top level object of `foo` with a property `bar`.

**Comment by tve** [comment]
Interesting, can you point to the documentation that mentions this?

**Comment by knolleary** [comment]
I don't think I can... but it is how it has always been. I will add something to https://nodered.org/docs/user-guide/context

**Comment by tve** [comment]
Yeah.... I've been using this for years and always do get-modify-set to update data structures, never occurred to me that the store allows direct updates down in the data structure...
The reason I hit this issue is that I tried to use node IDs in the key and sometimes node IDs contain dots...
